### DISCUSSION TITLE
Utility functions for NeoVim

### DISCUSSION THREAD

**Discussion by danielo515** [twpayne/chezmoi#2975] [Show and tell]
Hey guys! As a NeoVim and Chezmoi user, I've created a little Lua module that I'd like to share with you. This module contains some utility functions that allow you to execute Chezmoi from NeoVim, as well as re-add files that have been edited outside of the Chezmoi root from NeoVim.

Here's a brief overview of what each function does:

-  `executeChezmoi`: This function allows you to execute Chezmoi from Vim. Simply call this function and pass in the Chezmoi command you'd like to execute, and it will run in the terminal.

-  `re_add`: This function allows you to re-add files that have been edited outside of the Chezmoi root. Simply call this function and pass in the path to the file you'd like to re-add, and it will be added back to Chezmoi.

- `get_chezmoi_dir`: returns the root directory of chezmoi. Useful if you want to setup autocomands that only happen within the root directory chezmoi uses.

It also has  a little boolean flag to help you determine if chezmoi is available as an executable command or not. It requires `plenary`, but most of the plugins you are probably using already requie it.

```lua
local Job = require "plenary.job"
local has_chezmoi = vim.fn.executable "chezmoi" == 1
local M = { has_chezmoi = has_chezmoi }

function M.executeChezmoi(args)
  local result, _ = Job:new({
    command = "chezmoi",
    args = args,
    -- Do not set env prop or it will  fail
    on_exit = function(_, return_val)
      if return_val ~= 0 then
        print("Chezmoi command failed to execute: " .. table.concat(args, " "))
        return
      end
    end,
  }):sync()
  return result
end

function M.get_chezmoi_dir()
  local results = M.executeChezmoi { "source-path" }
  local path = results[1]
  return path
end

function M.re_add(path)
  local result, _ = Job:new({
    command = "chezmoi",
    args = { "managed", "-i", "files", path },
    -- Do not set env prop or it will  fail
    on_exit = function(_, return_val)
      if return_val ~= 0 then
        print "Chezmoi command failed to execute"
        return
      end
    end,
  }):sync()
  local isIncluded = #result > 0
  -- vim.pretty_print("file under control normal: ", isIncluded)
  -- vim.pretty_print(result, #result, isIncluded)
  if not isIncluded then
    return
  end
  vim.ui.select({ "re-add", "ignore" }, { prompt = "Chezmoi managed file edited:" .. path }, function(choice)
    local saidYes = choice == "re-add"
    if saidYes then
      local res, code = Job:new({ command = "chezmoi", args = { "add", "-f", path } }):sync()
    end
  end)
end


return M
```

The way I use this is generally through some autocommands that execute the different functions when needed. Here you  have one example:

```lua
local command = vim.api.nvim_create_autocmd
local chezmoi = require("user.chezmoi")
-- Re add to chezmoi snippet files, because that is the only way to have live reload of snippets
command("BufWritePost", {
  pattern = "*/.config/*",
  callback = function()
    local file = vim.fn.expand("%:p")
    chezmoi.re_add(file)
  end,
})
-- Apply chezmoi whenever a dotfile is updated
command("BufWritePost", {
  pattern = chezmoi.get_chezmoi_dir() .. "/*",
  callback = function()
    local file = vim.fn.expand("%:p")
    chezmoi.executeChezmoi({ "apply", "-v", "--refresh-externals=false", "--source-path", file })
  end,
})

```

**Comment by twpayne** [comment]
Very neat, thanks for sharing!

**Comment by kaddkaka** [comment]
What is different between `executeChezmoi` and plain old `:!` (`:!chezmoi`) in (n)vim?

**Reply by danielo515** [reply]
Depends on what you want to do with the output. executeChezmoi will collect the output and provide it to you as a result in a lua table, where the plain old will not. It all depends on your specific use case. Also, if you were to build the arguments programatically it is easier to build a lua table than concatenate strings, but as I said, depends.
